src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -282,21 +290,59 @@ public DeconstructionVariable(ArrayBuilder<DeconstructionVariable> variables, CS
         }
 
         /// <summary>
-        /// For cases where the RHS of a deconstruction-assignment has no type (TupleLiteral), we squint and look at the LHS as a tuple type to give the RHS a type.
+        /// For cases where the RHS of a deconstruction-declaration is a tuple literal, we merge type information from both the LHS and RHS.
+        /// For cases where the RHS of a deconstruction-assignment is a tuple literal, the type information from the LHS determines the merged type, since all variables have a type.
+        /// Returns null if a merged tuple type could not be fabricated.
         /// </summary>
-        private static TypeSymbol MakeTupleTypeFromDeconstructionLHS(ArrayBuilder<DeconstructionVariable> topLevelCheckedVariables, DiagnosticBag diagnostics, CSharpCompilation compilation)
+        private static TypeSymbol MakeMergedTupleType(ArrayBuilder<DeconstructionVariable> lhsVariables, BoundTupleLiteral rhsLiteral, CSharpSyntaxNode syntax, DiagnosticBag diagnostics, CSharpCompilation compilation)
         {
-            var typesBuilder = ArrayBuilder<TypeSymbol>.GetInstance(topLevelCheckedVariables.Count);
-            foreach (var variable in topLevelCheckedVariables)
+            int leftLength = lhsVariables.Count;
+            int rightLength = rhsLiteral.Arguments.Length;
+
+            var typesBuilder = ArrayBuilder<TypeSymbol>.GetInstance(lhsVariables.Count);
+            for (int i = 0; i < rightLength; i++)
             {
-                if (variable.HasNestedVariables)
-                {
-                    typesBuilder.Add(MakeTupleTypeFromDeconstructionLHS(variable.NestedVariables, diagnostics, compilation));
-                }
-                else
+                BoundExpression element = rhsLiteral.Arguments[i];
+                TypeSymbol mergedType = element.Type;
+
+                if (i < leftLength)
```

Comment:
What is the scenario where the right length is greater than left length? And are we reporting an error if there are `null` types in those elements?